### PR TITLE
Avoid recompilation issue in `JaxSimModelData` instantiation

### PR DESCRIPTION
**Pull request by flferretti**
This pull request includes several changes to the `JaxSimModelData` class in the `src/jaxsim/api/data.py` file. The main goal of these changes is to avoid recompilation issue in `JaxSimModelData` instantiation. Below is a summary of the most important changes:

Improvements to contact state handling:

* Modified the `contact_state` field to use a default factory, ensuring it is always initialized as an empty dictionary. (`src/jaxsim/api/data.py`, [src/jaxsim/api/data.pyL63-R63](diffhunk://#diff-d377d6d95ae5a2ffe62db73e1f1c4dbaf6658f1c57a340aa8d15d70e78d3dde2L63-R63))
* Updated the `build` method to use `contact_state.get` instead of `contact_state.setdefault` for initializing the "tangential_deformation" field avoiding in-place mutations. (`src/jaxsim/api/data.py`, [src/jaxsim/api/data.pyL177-R177](diffhunk://#diff-d377d6d95ae5a2ffe62db73e1f1c4dbaf6658f1c57a340aa8d15d70e78d3dde2L177-R177))
* Simplified the `replace` method by directly assigning a new dictionary to `contact_state` when dealing with soft contacts. (`src/jaxsim/api/data.py`, [src/jaxsim/api/data.pyL440-R439](diffhunk://#diff-d377d6d95ae5a2ffe62db73e1f1c4dbaf6658f1c57a340aa8d15d70e78d3dde2L440-R439))

Array reshaping and normalization:

* Changed the reshaping of `joint_positions`, `joint_velocities`, `base_quaternion`, and `base_position` to use `jnp.atleast_2d` to ensure they are always 2-dimensional arrays. (`src/jaxsim/api/data.py`, [src/jaxsim/api/data.pyL453-R460](diffhunk://#diff-d377d6d95ae5a2ffe62db73e1f1c4dbaf6658f1c57a340aa8d15d70e78d3dde2L453-R460))
* Removed the broadcasting of arrays and instead reshaped them directly to match the required dimensions in the `replace` method. (`src/jaxsim/api/data.py`, [src/jaxsim/api/data.pyL497-L517](diffhunk://#diff-d377d6d95ae5a2ffe62db73e1f1c4dbaf6658f1c57a340aa8d15d70e78d3dde2L497-L517))

----
📚 Documentation preview 📚: https://jaxsim--395.org.readthedocs.build//395/

### CODE REVIEW

**Comment by CarlottaSartore** [comment]
Thanks @flferretti, I did not have yet the chance to go through the PR but I was wondering  since the re compilation was introduces by a previous PR (I think is the batched data one) does this PR revert the introduce changes ?

**Comment by flferretti** [comment]
> Thanks @flferretti, I did not have yet the chance to go through the PR but I was wondering since the re compilation was introduces by a previous PR (I think is the batched data one) does this PR revert the introduce changes ?

Thanks for your message! This PR does not revert the changes introduced in the previous one but rather builds upon them. The details of how recompilation is handled are outlined in the changes and summed up in the PR description, so I’d recommend reviewing them first to get a clearer picture. Let me know if anything needs further clarification
